charmcraft/commands/store/__init__.py

```diff
@@ -722,6 +737,222 @@ def run(self, parsed_args):
+        elif output_bundle is not None:
+            if not output_bundle.suffix:
+                output_bundle /= "bundle.yaml"
+            for parent in output_bundle.parents:
+                if parent.exists():
+                    if os.access(parent, os.W_OK):
+                        break
+                    raise CraftError(f"Bundle output directory not writable: {str(parent)}")
+
+        # Load bundle
+        bundle_path = self.config.project.dirpath / "bundle.yaml"
+        bundle_config = load_yaml(bundle_path)
+        if bundle_config is None:
+            raise CraftError(f"Missing or invalid main bundle file: {(str(bundle_path))}")
+        bundle_name = bundle_config.get("name")
+        if not bundle_name:
+            raise CraftError(
+                "Invalid bundle config; missing a 'name' field indicating the bundle's name in "
+                f"file {str(bundle_path)!r}."
+            )
+        emit.progress("Determining charms to promote")
+        charms = [c["charm"] for c in bundle_config["applications"].values()]
+        errant_excludes = []
+        for excluded in parsed_args.exclude:
+            try:
+                charms.remove(excluded)
+            except ValueError:
+                errant_excludes.append(excluded)
+        if errant_excludes:
+            bad_charms = utils.humanize_list(errant_excludes, "and")
+            raise CraftError(
+                f"Bundle does not contain the following excluded charms: {bad_charms}"
+            )
+
+        store = Store(self.config.charmhub)
+        registered_names: List[Entity] = store.list_registered_names(include_collaborations=True)
+        name_map = {name.name: name for name in registered_names}
+
+        if bundle_name not in name_map:
+            raise CraftError(
+                f"Cannot modify bundle {bundle_name}. Ensure the bundle exists and that you have "
+                "been made a collaborator."
+            )
+        elif name_map[bundle_name].entity_type != EntityType.bundle:
+            entity_type = name_map[bundle_name].entity_type
+            raise CraftError(f"Store Entity {bundle_name} is a {entity_type}, not a bundle.")
+
+        invalid_charms = []
+        non_charms = []
+        for charm_name in charms:
+            if charm_name not in name_map:
+                invalid_charms.append(charm_name)
+            elif name_map[charm_name].entity_type != EntityType.charm:
+                non_charms.append(charm_name)
+        if invalid_charms:
+            charm_list = utils.humanize_list(invalid_charms, "and")
+            raise CraftError(
+                "The following entities do not exist or you are not a collaborator on them: "
+                f"{charm_list}"
+            )
+        if non_charms:
+            non_charm_list = utils.humanize_list(non_charms, "and")
+            raise CraftError(f"The following store entities are not charms: {non_charm_list}")
```

Comment:
Does the API able to tell the charm exists or not?
I think something like:

The following charms you own/collaboration will be promoted:
xxxx: beta -> xxxx: stable
xxxx: beta -> xxxx: stable

The following charms you do not own/collaboration will be ignored:
xxxx: beta

So we don't need to ask user to exclude everything.

Reply:
The API we're using won't reveal any charms you can't write to, which is why we can't distinguish charms the user doesn't own or isn't a collaborator on from charms that don't exist. We could retrieve those charms the same way the website does, but I don't think that's a stable public API. If we want to do that we'll probably need to get the store team to add an endpoint to the developer API. The `exclude` option still needs to exist for excluding promotions on charms the user already owns, but we could detect invalid charms and simply exclude them instead. 

For the MVP, I opted for maximum explicitness after discussing it with Facundo. This particular behaviour is undefined in the spec, so we'll probably want to discuss the options with stakeholders. The way I'd imagine we'd implement it would be with a pair of arguments (`--auto-exclude` and `--no-auto-exclude`) to prefer an answer, defaulting to either asking when interactive or failing (equivalent to `--no-auto-exclude`) when not interactive.

Reply:
(Marking this as resolved for now, but it might be something to discuss in a further design meeting)